stream: improving error handling

mafintosh opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
stream

This PR improves error handling for streams in a few ways.

1. It ensures that no user defined methods (_read, _write, ...) are run *after* .destroy has been called.
2. It introduces an explicit error to tell the user if they are write to write, etc to the stream *after* it has been destroyed.
3. It makes streams *always* emit `close` as the last thing after they have been destroyed
4. Changes the default _destroy to *not* gracefully end streams.

Especially 4. makes it easier for userland modules to handle stream errors as they don't appear as graceful closes anymore. Seen issues being with end-of-stream/pump because of this.

mafintosh commented:
/cc @mcollina @nodejs/streams

vsemozhetbyt commented on doc/api/errors.md:
```diff
@@ -1505,6 +1505,12 @@ An unspecified or non-specific system error has occurred within the Node.js
+<a id="ERR_STREAM_DESTROYED"></a>
+### ERR_STREAM_DESTROYED
+
+An stream method was called that cannot complete cause the stream was been
```
Nits:
An stream -> A stream
was been -> has been

vsemozhetbyt commented on lib/_stream_readable.js:
```diff
@@ -106,6 +106,9 @@ function ReadableState(options, stream) {
   this.readableListening = false;
   this.resumeScheduled = false;
 
+  // True if the close was already emitted and should not emitted again
```
A nit: should not emitted -> should not be emitted

vsemozhetbyt commented on lib/_stream_writable.js:
```diff
@@ -134,6 +134,9 @@ function WritableState(options, stream) {
   // True if the error was already emitted and should not be thrown again
   this.errorEmitted = false;
 
+  // True if the close was already emitted and should not emitted again
```
A nit: should not emitted -> should not be emitted

jasnell commented:
@nodejs/http2 ... just a quick note that this will have a relatively small impact on the `'close'` event currently emitted for `Http2Stream` and `Http2Session`.

mafintosh commented:
Added docs about the `close` event and the `ERR_STREAM_DESTROYED` error.

mcollina commented on lib/internal/http2/core.js:
```diff
@@ -1471,6 +1471,11 @@ class Http2Stream extends Duplex {
+    // Set close emitted, so the stream destruction does not
+    // emit them
+    this._readableState.closeEmitted = true;
+    this._writableState.closeEmitted = true;
```
I'm not very fond of this lines. A current goal is to reduce (or totally remove) the number of access to `_readableState` and `_writableState`. However, this requires it. Maybe emitting  `'close'` automatically can be moved to an option?

mafintosh replied:
This is only because http2 currently emits a 'code' argument with the close event. The idea is to get rid of that in http2 and move the 'code' to a property down the line, but this ensures backwards compat right now.

Should I add a comment explaining that?

mcollina replied:
What I mean is that in both `fs` and `http2` for compat reason we need to not emit `'close'` on destroy. I think this might actually be a legit use case, and so we should support it with an option at construction.

mafintosh replied:
The `fs` thing is legit, but we should just get rid of the code in `http2` imo, but that's another discussion. I like the idea of an option, adding it.

mcollina commented on lib/_stream_readable.js:
```diff
@@ -437,7 +441,8 @@ Readable.prototype.read = function(n) {
       state.needReadable = true;
     // call internal read method
-    this._read(state.highWaterMark);
+    if (!state.destroyed)
```
I think this check can me moved in the previous `if` block: `state.ended || state.reading || state.destroyed`.

mcollina commented on lib/_stream_transform.js:
```diff
@@ -132,6 +132,8 @@ function Transform(options) {
 
 function prefinish() {
+  if (this._readableState.destroyed)
+    return;
```
can you add one more test for the `_flush()` logic and `destroy()`? It seems you are fixing a bug here. Or without this change some tests would not pass anymore?

mafintosh commented:
@mcollina fixed your comments

mcollina commented on lib/net.js:
```diff
@@ -219,6 +226,11 @@ function Socket(options) {
   else if (options === undefined)
     options = {};
+  else
+    options = copyObject(options);
```
You can use `Object.assign`.

mafintosh commented:
@mcollina nit fixed

mcollina commented:
@mafintosh can you add `emitClose`  to the documented options?

mafintosh commented:
@mcollina fixed the docs

Trott commented on doc/api/errors.md:
```diff
@@ -1505,6 +1505,12 @@ An unspecified or non-specific system error has occurred within the Node.js
+<a id="ERR_STREAM_DESTROYED"></a>
+### ERR_STREAM_DESTROYED
+
+A stream method was called that cannot complete cause the stream was been
```
Seems like `cause` should be `because` here, I think?

jasnell replied:
s/cause/because

mafintosh commented:
@jasnell @Trott fixed

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/12879/
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1250/

mcollina commented:
CI (Again): https://ci.nodejs.org/job/node-test-pull-request/12881/

mafintosh commented:
@mcollina ci seemed to have an issue?

mcollina commented:
We are testing in so many system that a botched run happens every now and then.

addaleax commented on doc/api/stream.md:
```diff
@@ -1378,6 +1381,8 @@ constructor and implement the `writable._write()` method. The
     `Buffer` or `Uint8Array` if supported by the stream implementation.
     Defaults to `false`
+  * `emitClose` {boolean} Whether or not the stream should emit `close`
+    after it has been destroyed. Defaults to `true`
```
This should be documented with a `changes:` YAML changelog entry for this heading (grep around for that string in the docs to see how it’s used)

addaleax commented on test/parallel/test-zlib-write-after-close.js:
```diff
@@ -29,9 +29,9 @@ zlib.gzip('hello', common.mustCall(function(err, out) {
   common.expectsError(
     () => unzip.write(out),
     {
-      code: 'ERR_ZLIB_BINDING_CLOSED',
```
Can this error still be triggered at all? I don’t see how, so it’s probably best to remove it from the errors list, from the docs, and turn the handle check in `lib/zlib.js` into an assertion.

cjihrig commented on doc/api/stream.md:
```diff
@@ -535,8 +535,10 @@ added: v8.0.0
 
-Destroy the stream, and emit the passed error. After this call, the
-writable stream has ended. Implementors should not override this method,
+Destroy the stream, and emit the passed error and a close event.
```
Should error and close be in backticks? It looks like they are, in the next section.

BridgeAR commented on lib/net.js:
```diff
@@ -219,6 +219,11 @@ function Socket(options) {
   else if (options === undefined)
     options = {};
+  else
+    options = Object.assign({}, options);
```
Please use `util._extends`. It is still faster than `Object.assign`.

BridgeAR commented on lib/net.js:
```diff
@@ -219,6 +219,11 @@ function Socket(options) {
+  else
+    options = Object.assign({}, options);
+
+  // for backwards compat do not emit close on destroy.
```
Nit: please use caps for the the char of a comment.

BridgeAR commented on lib/internal/errors.js:
```diff
@@ -633,6 +633,7 @@ E('ERR_SOCKET_DGRAM_NOT_RUNNING', 'Not running');
 E('ERR_STDERR_CLOSE', 'process.stderr cannot be closed');
 E('ERR_STDOUT_CLOSE', 'process.stdout cannot be closed');
 E('ERR_STREAM_CANNOT_PIPE', 'Cannot pipe, not readable');
+E('ERR_STREAM_DESTROYED', 'stream.destroy() was called');
```
The description is not that descriptive as far as I see it. Would you be so kind and try to reword?

mafintosh replied:
We don't have much context to why it was destroyed other than the fact that someone called the destroy method. Would something like `'Stream is destroyed'` be better?

BridgeAR replied:
Hm... what about: `The stream is already destroyed. Further operations are not possible anymore.` (I am not that happy with that either but it is maybe a bit more expressive?). Instead of `Further operations...` it would also be possible to say e.g.: `Calling ${functionName} is not possible anymore.`

mafintosh commented:
Rebasing against master

mafintosh commented:
@addaleax @BridgeAR fixed your comments

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/13175/

lpinca commented on doc/api/stream.md:
```diff
@@ -1158,8 +1160,9 @@ myReader.on('readable', () => {
-readable stream will release any internal resources.
+Destroy the stream, and emit `'error'` and `close`. After this call, the
+readable stream will release any internal resources and subsequent calls
+to `read` will result in an `ERR_STREAM_DESTROYED` error.
```
I have code that rely on the current behavior (`read` after destroy) to ensure that all buffered data is written out.

```js
socket.pipe(parser);
socket.on('close', () => {
  // Ensure that all buffered data is written to the parser if the socket is
  // destroyed due to a network error (ECONNRESET, EPIPE, etc.) and the `'end'`
  // event is not emitted.
  socket.read();
  parser.end();
});
```

I guess that with this change it will no longer work, how can I fix that?

mafintosh replied:
This is actually a typo. It should be '.push' and now `.read'. Fixed it just now. Also you are right that you should be able to get data out of the buffer after .destroy(). That is also fixed.

lpinca replied:
Awesome, thanks.

mcollina commented:
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1292/

mcollina commented:
I fear this might be breaking Koa, let's have another run.

```
undefinedadded 473 packages in 45.641s
 > koa@2.5.0 test /Users/iojs/build/workspace/citgm-smoker/nodes/osx1010/citgm_tmp/b6e916d2-2dc9-4087-9795-b17ed8174197/koa
 > jest
  PASS  test/application/respond.js
  PASS  test/context/onerror.js
 Thu, 15 Feb 2018 11:56:31 GMT http-errors deprecated non-first-argument status code; replace with createError(422, ...) at lib/context.js:93:11
 Thu, 15 Feb 2018 11:56:31 GMT http-errors deprecated non-error status code; use only 4xx or 5xx status codes at lib/context.js:93:11
  PASS  test/context/throw.js
  PASS  test/response/body.js
 events.js:122
     throw err;
     ^
 Error [ERR_UNHANDLED_ERROR]: Unhandled error. (Error: non-error thrown: {})
     at Application.emit (events.js:120:17)
     at Object.onerror (/Users/iojs/build/workspace/citgm-smoker/nodes/osx1010/citgm_tmp/b6e916d2-2dc9-4087-9795-b17ed8174197/koa/lib/context.js:117:14)
     at ReadStream.err (/Users/iojs/build/workspace/citgm-smoker/nodes/osx1010/citgm_tmp/b6e916d2-2dc9-4087-9795-b17ed8174197/koa/lib/response.js:169:47)
     at ReadStream.emit (events.js:131:13)
     at readableAddChunk (_stream_readable.js:242:16)
     at ReadStream.Readable.push (_stream_readable.js:207:10)
     at fs.read (fs.js:2334:12)
     at callback (/Users/iojs/build/workspace/citgm-smoker/nodes/osx1010/citgm_tmp/b6e916d2-2dc9-4087-9795-b17ed8174197/koa/node_modules/graceful-fs/polyfills.js:128:19)
     at FSReqWrap.wrapper [as oncomplete] (fs.js:816:17)
 npm ERR! Test failed.  See above for more details.
```

CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1295/
CITGM (master): https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1296/

mafintosh commented:
I cannot reproduce that koa error locally

mafintosh commented:
Tried reproducing the koa failure on a debian machine as well, with no luck

BridgeAR commented:
I can not reproduce it locally either but it clearly shows up on CITGM.

ronkorving commented:
Is that really an empty object being thrown, or are we not outputting the object's properties? I assume it's the latter, but that's quite a time-waster in trying to figure out what's going wrong.

BridgeAR commented:
Hm, I am not really sure what to do. Without being able to reproduce the issue it is hard to debug this. Does anyone has a idea?

mafintosh commented:
@BridgeAR @mcollina Running the koa tests a 1000 times seem to once in a while reproduce it for me. Seeing if the failure is related...

gibfahn commented:
>@BridgeAR @mcollina Running the koa tests a 1000 times seem to once in a while reproduce it for me. Seeing if the failure is related...

>I can not reproduce it locally either but it clearly shows up on CITGM.

CitGM runs all the module test suites in parallel, so the machines are generally pretty heavily loaded. Adding load might make it easier to reproduce.

mafintosh commented:
Rebased master and fixed koa (locally at least) by not having .push throw after destroy but instead just not queuing data. 

This makes sense as well as push is usually called async in the _read function where the stream can be destroyed in between.

mafintosh commented:
@jasnell or @mcollina could you trigger another citgm run? it should pass now :)

mafintosh commented:
or @BridgeAR o/

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/13510/

addaleax commented on doc/api/stream.md:
```diff
@@ -1382,6 +1385,12 @@ constructor and implement the `writable._write()` method. The
+changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/18438
+    description: add emitClose option to specify if close is emitted on destroy
```
Tiny nit: I think so far all of these desciprions are capitalized & punctuated, and it would be nice to keep in line with that. Also, putting `close` in a code block via backticks might be nice, to distinguish it as an event name.

mcollina commented:
Landed in a5964eb

joyeecheung commented:
This is breaking the linter on master

```
07:52:11 doc/api/stream.md
07:52:11   1392:84  warning  Line must be at most 80 characters  maximum-line-length  remark-lint
```

https://ci.nodejs.org/job/node-test-linter/16731/console
